From 4d02f4c0b576e331e5756120e85ad25454140be9 Mon Sep 17 00:00:00 2001
From: Peter M <petermm@gmail.com>
Date: Thu, 15 Oct 2026 11:54:06 +0200
Subject: [PATCH] bsps/riscv: Add ESP32-C6 BSP (riscv/esp32c6db)

Add ESP32-C6 board support package (riscv/esp32c6db) with rv32imac
architecture and ilp32 ABI.

- Map unified 512 KiB HP SRAM (0x40800000 - 0x40880000)
- Set direct-boot signature header and flash window at 0x42000000
- Reserve 0x16000 (88 KiB) at the top of HP SRAM for ROM data, BSS, and stack
- Add ESP32-C6 interrupt matrix and INTPRI routing with 77 sources
- Adapt systimer, GPIO matrix, SPI, I2C, and watchdog registers
- Enable BSP_START_COPY_SECTIONS for flash data copying at boot

Signed-off-by: Peter M <petermm@gmail.com>
---
 bsps/riscv/esp32/bt/bt_c3.c                   |  28 +++
 bsps/riscv/esp32/i2c/i2c_c3.c                 |   6 +-
 bsps/riscv/esp32/include/bsp.h                |   7 +
 bsps/riscv/esp32/include/bsp/pin.h            |   6 +
 .../riscv/esp32/include/c6/chip_definitions.h | 180 +++++++++++++
 bsps/riscv/esp32/irq/irq_c3.c                 | 237 +++++++++++++++---
 bsps/riscv/esp32/spi/spi_c3.c                 |   6 +
 bsps/riscv/esp32/start/bspstart.c             |  19 +-
 spec/build/bsps/riscv/esp32/abi.yml           |   9 +
 spec/build/bsps/riscv/esp32/bspesp32c6db.yml  |  23 ++
 spec/build/bsps/riscv/esp32/grp.yml           |  10 +
 spec/build/bsps/riscv/esp32/linkcmds.yml      |  23 +-
 .../build/bsps/riscv/esp32/optchipvariant.yml |   3 +
 .../riscv/esp32/optdataflashmappedbase.yml    |  23 ++
 .../bsps/riscv/esp32/optdramromreserve.yml    |   2 +
 spec/build/bsps/riscv/esp32/optiramdelta.yml  |  23 ++
 .../bsps/riscv/esp32/optramcodeorigin.yml     |  23 ++
 spec/build/bsps/riscv/esp32/optramorigin.yml  |  23 ++
 spec/build/bsps/riscv/esp32/optramsize.yml    |  23 ++
 spec/build/bsps/riscv/esp32/optvariant.yml    |   3 +
 .../bsps/riscv/optbspstartcopysections.yml    |   4 +-
 .../bsps/riscv/optespressifdirectboot.yml     |   4 +-
 22 files changed, 627 insertions(+), 58 deletions(-)
 create mode 100644 bsps/riscv/esp32/include/c6/chip_definitions.h
 create mode 100644 spec/build/bsps/riscv/esp32/bspesp32c6db.yml
 create mode 100644 spec/build/bsps/riscv/esp32/optdataflashmappedbase.yml
 create mode 100644 spec/build/bsps/riscv/esp32/optiramdelta.yml
 create mode 100644 spec/build/bsps/riscv/esp32/optramcodeorigin.yml
 create mode 100644 spec/build/bsps/riscv/esp32/optramorigin.yml
 create mode 100644 spec/build/bsps/riscv/esp32/optramsize.yml

diff --git a/bsps/riscv/esp32/bt/bt_c3.c b/bsps/riscv/esp32/bt/bt_c3.c
index 156cf3a..1271e55 100644
--- a/bsps/riscv/esp32/bt/bt_c3.c
+++ b/bsps/riscv/esp32/bt/bt_c3.c
@@ -40,6 +40,32 @@
 
 #include <rtems/counter.h>
 
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+
+void bsp_esp32_bt_get_state( bsp_esp32_bt_state *state )
+{
+  state->dig_pwc = 0;
+  state->dig_iso = 0;
+  state->clk_en = 0;
+}
+
+bool bsp_esp32_bt_is_enabled( void )
+{
+  return false;
+}
+
+rtems_status_code bsp_esp32_bt_enable( void )
+{
+  return RTEMS_UNSATISFIED;
+}
+
+rtems_status_code bsp_esp32_bt_disable( void )
+{
+  return RTEMS_SUCCESSFUL;
+}
+
+#else
+
 #define RTC_REG( off ) \
   ( *( (volatile uint32_t *) ( RTC_CNTL_BASE + (off) ) ) )
 #define SYSCON_REG( off ) \
@@ -209,3 +235,5 @@ rtems_status_code bsp_esp32_bt_disable( void )
 
   return RTEMS_SUCCESSFUL;
 }
+
+#endif /* !defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H) */
diff --git a/bsps/riscv/esp32/i2c/i2c_c3.c b/bsps/riscv/esp32/i2c/i2c_c3.c
index 821febf..f9630e0 100644
--- a/bsps/riscv/esp32/i2c/i2c_c3.c
+++ b/bsps/riscv/esp32/i2c/i2c_c3.c
@@ -162,9 +162,13 @@
 #define IO_MUX_MCU_SEL_MASK   ( 0x7U << IO_MUX_MCU_SEL_SHIFT )
 #define IO_MUX_FUNC_GPIO      1
 
-/* GPIO matrix signal numbers for the one I2C controller this chip has. */
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+#define I2C_SIG_SCL 45
+#define I2C_SIG_SDA 46
+#else
 #define I2C_SIG_SCL 53
 #define I2C_SIG_SDA 54
+#endif
 
 /* Where the bus is on the ESP32-C3-DevKitM-1, and what esp-idf's examples
  * use.  A board that wires it elsewhere wants these changed; they are here
diff --git a/bsps/riscv/esp32/include/bsp.h b/bsps/riscv/esp32/include/bsp.h
index 99bcdbd..88c1ac8 100644
--- a/bsps/riscv/esp32/include/bsp.h
+++ b/bsps/riscv/esp32/include/bsp.h
@@ -53,6 +53,9 @@
 
 #include <bspopts.h>
 #include <bsp/default-initial-extension.h>
+#include <rtems/score/basedefs.h>
+
+#include RTEMS_XSTRING( ESPRESSIF_CHIP_HEADER )
 
 #include <rtems/devnull.h>
 
@@ -61,8 +64,12 @@
  * brought out to pads, GPIO0 to GPIO21.  Both numbers describe the C3; a
  * second chip in this family would move them to its chip definitions header.
  */
+#ifndef BSP_GPIO_PIN_COUNT
 #define BSP_GPIO_PIN_COUNT     22
+#endif
+#ifndef BSP_GPIO_PINS_PER_BANK
 #define BSP_GPIO_PINS_PER_BANK 32
+#endif
 
 #ifdef __cplusplus
 extern "C" {
diff --git a/bsps/riscv/esp32/include/bsp/pin.h b/bsps/riscv/esp32/include/bsp/pin.h
index 51f2fce..e8d777f 100644
--- a/bsps/riscv/esp32/include/bsp/pin.h
+++ b/bsps/riscv/esp32/include/bsp/pin.h
@@ -46,7 +46,13 @@ extern "C" {
 /**
  * @brief The number of pads this part has, GPIO0 through GPIO21.
  */
+#ifndef BSP_PIN_COUNT
+#if defined(BSP_GPIO_PIN_COUNT)
+#define BSP_PIN_COUNT BSP_GPIO_PIN_COUNT
+#else
 #define BSP_PIN_COUNT 22
+#endif
+#endif
 
 /**
  * @brief Claims @a pin for @a owner.
diff --git a/bsps/riscv/esp32/include/c6/chip_definitions.h b/bsps/riscv/esp32/include/c6/chip_definitions.h
new file mode 100644
index 0000000..43d8e16
--- /dev/null
+++ b/bsps/riscv/esp32/include/c6/chip_definitions.h
@@ -0,0 +1,180 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+
+/**
+ * @file
+ *
+ * @ingroup ESP32
+ *
+ * @brief Interrupt and address definitions specific to ESP32-C6.
+ */
+
+/*
+ * Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
+ * Copyright (C) 2026 Peter M. <petermm@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H
+#define LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H
+
+#ifndef ASM
+
+#include <stdint.h>
+
+#endif /* ASM */
+
+/*
+ * 1-31 for the internal VIC with IRQ 0 being invalid since it corresponds to
+ * machine exception. This controller is fronted by an interrupt matrix
+ * supporting 77 peripheral interrupts (0-76). The count of 78 below includes
+ * 77 peripheral interrupts plus invalid interrupt 0.
+ */
+#define RISCV_MAXIMUM_EXTERNAL_INTERRUPTS 78
+#define BSP_INTERRUPT_VECTOR_COUNT        RISCV_MAXIMUM_EXTERNAL_INTERRUPTS
+
+#define BSP_INTERRUPT_CUSTOM_VALID_VECTOR
+
+#define BSP_GPIO_PIN_COUNT     31
+#define BSP_GPIO_PINS_PER_BANK 32
+
+/*
+ * The WiFi sources, which are peripheral interrupts 0 to 3 on this part.
+ */
+#define WIFI_MAC_INTR          0
+#define WIFI_MAC_NMI_INTR      1
+#define WIFI_PWR_INTR          2
+#define WIFI_BB_INTR           3
+
+/*
+ * Bluetooth and Low Power sources.
+ */
+#define BT_MAC_INTR            4
+#define BT_BB_INTR             5
+#define BT_BB_NMI_INTR         6
+#define LP_TIMER_INTR          7
+#define COEX_INTR              8
+#define BLE_TIMER_INTR         9
+#define BLE_SEC_INTR          10
+#define I2C_MASTER_INTR       11
+#define ZB_MAC_INTR           12
+#define PMU_INTR              13
+#define EFUSE_INTR            14
+#define LP_RTC_TIMER_INTR     15
+#define LP_UART_INTR          16
+#define LP_I2C_INTR           17
+#define LP_WDT_INTR           18
+#define LP_PERI_TIMEOUT_INTR  19
+#define LP_APM_M0_INTR        20
+#define LP_APM_M1_INTR        21
+
+/*
+ * CPU software interrupts.
+ */
+#define SW_INTR_0             22
+#define SW_INTR_1             23
+#define SW_INTR_2             24
+#define SW_INTR_3             25
+
+#define ASSIST_DEBUG_INTR     26
+#define TRACE_INTR            27
+#define CACHE_INTR            28
+#define CPU_PERI_TIMEOUT_INTR 29
+#define GPIO_INTR             30
+#define GPIO_PROCPU_INTR      30
+#define GPIO_NMI_INTR         31
+#define PAU_INTR              32
+#define HP_PERI_TIMEOUT_INTR  33
+#define MODEM_PERI_TIMEOUT_INTR 34
+#define HP_APM_M0_INTR        35
+#define HP_APM_M1_INTR        36
+#define HP_APM_M2_INTR        37
+#define HP_APM_M3_INTR        38
+#define LP_APM0_INTR          39
+#define MSPI_INTR             40
+#define I2S0_INTR             41
+#define I2S_INTR              41
+#define UHCI0_INTR            42
+#define UART_INTR             43
+#define UART1_INTR            44
+#define LEDC_INTR             45
+#define TWAI0_INTR            46
+#define TWAI_INTR             46
+#define TWAI1_INTR            47
+#define USB_SERIAL_JTAG_INTR  48
+#define RMT_INTR              49
+#define I2C_EXT0_INTR         50
+#define TG_T0_INTR            51
+#define TG0_T1_INTR           52
+#define TG_WDT_INTR           53
+#define TG1_T0_INTR           54
+#define TG1_T1_INTR           55
+#define TG1_WDT_INTR          56
+#define SYSTIMER_TARGET0_INTR 57
+#define SYSTIMER_TARGET1_INTR 58
+#define SYSTIMER_TARGET2_INTR 59
+#define VDIGTAL_ADC_INTR      60
+#define APB_ADC_INTR          60
+#define MCPWM0_INTR           61
+#define PCNT_INTR             62
+#define PARL_IO_INTR          63
+#define SLC0_INTR             64
+#define SLC_INTR              65
+#define GDMA_CH0_INTR         66
+#define GDMA_CH1_INTR         67
+#define GDMA_CH2_INTR         68
+#define GDMA_OUT_CH0_INTR     69
+#define GDMA_OUT_CH1_INTR     70
+#define GDMA_OUT_CH2_INTR     71
+#define GPSPI2_INTR           72
+#define GPSPI2_INTR_2         72
+#define AES_INTR              73
+#define SHA_INTR              74
+#define RSA_INTR              75
+#define ECC_INTR              76
+
+#define RISCV_INTERRUPT_VECTOR_SOFTWARE SW_INTR_0
+#define RISCV_INTERRUPT_VECTOR_TIMER    SYSTIMER_TARGET0_INTR
+
+#define GPIO_BASE            ( (uintptr_t) 0x60091000U )
+#define I2C_BASE             ( (uintptr_t) 0x60004000U )
+#define UART0_BASE           ( (uintptr_t) 0x60000000U )
+#define UART1_BASE           ( (uintptr_t) 0x60001000U )
+#define IO_MUX_BASE          ( (uintptr_t) 0x60090000U )
+#define SYSTIMER_BASE        ( (uintptr_t) 0x6000A000U )
+#define USB_SERIAL_JTAG_BASE ( (uintptr_t) 0x6000F000U )
+#define RTC_CNTL_BASE        ( (uintptr_t) 0x600B1000U )
+#define LP_AON_BASE          ( (uintptr_t) 0x600B1000U )
+#define LP_WDT_BASE          ( (uintptr_t) 0x600B1C00U )
+#define TIMG_BASE            ( (uintptr_t) 0x60008000U )
+#define INT_MATRIX_BASE      ( (uintptr_t) 0x60010000U )
+#define INTPRI_BASE          ( (uintptr_t) 0x600C5000U )
+#define SYSREG_BASE          ( (uintptr_t) 0x60095000U )
+#define PCR_BASE             ( (uintptr_t) 0x60096000U )
+
+/* ESP32-C6 ROM function addresses */
+#define UART_TX_ONE_CHAR_ADDR ( (void *) (uintptr_t) 0x40000058 )
+#define UART_RX_ONE_CHAR_ADDR ( (void *) (uintptr_t) 0x40000060 )
+#define UART_TX_FLUSH_ADDR    ( (void *) (uintptr_t) 0x40000074 )
+#define GPIO_OUTPUT_SET_ADDR  ( (void *) (uintptr_t) 0x400006e8 )
+
+#endif /* LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H */
diff --git a/bsps/riscv/esp32/irq/irq_c3.c b/bsps/riscv/esp32/irq/irq_c3.c
index 572c550..268cfc7 100644
--- a/bsps/riscv/esp32/irq/irq_c3.c
+++ b/bsps/riscv/esp32/irq/irq_c3.c
@@ -46,6 +46,48 @@ typedef struct peripheral_irq_map {
   uint8_t             cpu_int;
 } peripheral_irq_map_t;
 
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+static peripheral_irq_map_t irq_mappings[] = {
+  { .peripheral_int = WIFI_MAC_INTR, .cpu_int = 1 },
+  { .peripheral_int = WIFI_PWR_INTR, .cpu_int = 1 },
+  { .peripheral_int = WIFI_BB_INTR, .cpu_int = 1 },
+  { .peripheral_int = UHCI0_INTR, .cpu_int = 1 },
+  { .peripheral_int = SW_INTR_0, .cpu_int = 1 },
+  { .peripheral_int = SW_INTR_1, .cpu_int = 1 },
+  { .peripheral_int = SW_INTR_2, .cpu_int = 1 },
+  { .peripheral_int = SW_INTR_3, .cpu_int = 1 },
+  { .peripheral_int = GPIO_PROCPU_INTR, .cpu_int = 2 },
+  { .peripheral_int = GPSPI2_INTR_2, .cpu_int = 3 },
+  { .peripheral_int = I2S_INTR, .cpu_int = 4 },
+  { .peripheral_int = UART_INTR, .cpu_int = 5 },
+  { .peripheral_int = UART1_INTR, .cpu_int = 6 },
+  { .peripheral_int = LEDC_INTR, .cpu_int = 7 },
+  { .peripheral_int = EFUSE_INTR, .cpu_int = 7 },
+  { .peripheral_int = TWAI0_INTR, .cpu_int = 8 },
+  { .peripheral_int = TWAI1_INTR, .cpu_int = 8 },
+  { .peripheral_int = USB_SERIAL_JTAG_INTR, .cpu_int = 9 },
+  { .peripheral_int = LP_RTC_TIMER_INTR, .cpu_int = 10 },
+  { .peripheral_int = LP_WDT_INTR, .cpu_int = 10 },
+  { .peripheral_int = RMT_INTR, .cpu_int = 11 },
+  { .peripheral_int = I2C_EXT0_INTR, .cpu_int = 12 },
+  { .peripheral_int = I2C_MASTER_INTR, .cpu_int = 12 },
+  { .peripheral_int = TG_T0_INTR, .cpu_int = 13 },
+  { .peripheral_int = TG_WDT_INTR, .cpu_int = 14 },
+  { .peripheral_int = TG1_T0_INTR, .cpu_int = 15 },
+  { .peripheral_int = TG1_WDT_INTR, .cpu_int = 14 },
+  { .peripheral_int = SYSTIMER_TARGET0_INTR, .cpu_int = 16 },
+  { .peripheral_int = SYSTIMER_TARGET1_INTR, .cpu_int = 17 },
+  { .peripheral_int = SYSTIMER_TARGET2_INTR, .cpu_int = 18 },
+  { .peripheral_int = APB_ADC_INTR, .cpu_int = 19 },
+  { .peripheral_int = GDMA_CH0_INTR, .cpu_int = 20 },
+  { .peripheral_int = GDMA_CH1_INTR, .cpu_int = 21 },
+  { .peripheral_int = GDMA_CH2_INTR, .cpu_int = 22 },
+  { .peripheral_int = RSA_INTR, .cpu_int = 23 },
+  { .peripheral_int = AES_INTR, .cpu_int = 24 },
+  { .peripheral_int = SHA_INTR, .cpu_int = 25 },
+  { .peripheral_int = ASSIST_DEBUG_INTR, .cpu_int = 26 }
+};
+#else
 static peripheral_irq_map_t irq_mappings[] = {
   /*
    * WiFi, grouped onto channel 1 with the software interrupts.
@@ -128,9 +170,52 @@ static peripheral_irq_map_t irq_mappings[] = {
   { .peripheral_int = PMS_PERI_VIO_INTR, .cpu_int = 30 },
   { .peripheral_int = PMS_PERI_VIO_SIZE_INTR, .cpu_int = 31 }
 };
+#endif
 
 /* mapping registers are indexed directly from the base address */
 
+#define MATRIX_REG( reg ) *( (volatile uint32_t *) ( INT_MATRIX_BASE + reg ) )
+
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+/* ESP32-C6 Interrupt Matrix status registers */
+#define INTERRUPT_CORE0_INTR_STATUS_0_REG 0x134
+#define INTERRUPT_CORE0_INTR_STATUS_1_REG 0x138
+#define INTERRUPT_CORE0_INTR_STATUS_2_REG 0x13c
+
+#define INTPRI_REG( reg ) *( (volatile uint32_t *) ( INTPRI_BASE + ( reg ) ) )
+#define INTPRI_CPU_REG( reg ) *( (volatile uint32_t *) ( 0x20001000U + ( reg ) ) )
+
+static inline void esp32_cpu_int_enable_all( void )
+{
+  INTPRI_REG( 0x00 ) = 0xffffffffU;
+  INTPRI_CPU_REG( 0x00 ) = 0xffffffffU;
+}
+
+static inline void esp32_cpu_int_type_level( void )
+{
+  INTPRI_REG( 0x04 ) = 0x0U;
+  INTPRI_CPU_REG( 0x04 ) = 0x0U;
+}
+
+static inline void esp32_cpu_int_clear( uint8_t cpu_vector )
+{
+  INTPRI_REG( 0xa8 ) = 1U << cpu_vector;
+  INTPRI_REG( 0xa8 ) = 0U;
+  INTPRI_CPU_REG( 0xa8 ) = 1U << cpu_vector;
+  INTPRI_CPU_REG( 0xa8 ) = 0U;
+}
+
+static inline void esp32_cpu_int_set_prio( uint8_t cpu_int, uint32_t prio )
+{
+  INTPRI_REG( 0x0c + cpu_int * 4 ) = prio;
+  INTPRI_CPU_REG( 0x0c + cpu_int * 4 ) = prio;
+}
+
+static inline uint32_t esp32_cpu_int_get_prio( uint8_t cpu_int )
+{
+  return INTPRI_REG( 0x0c + cpu_int * 4 );
+}
+#else
 /* Bits here correspond directly to peripheral interrupt numbers */
 /* low 32 bits */
 #define INTERRUPT_CORE0_INTR_STATUS_0_REG 0xf8
@@ -148,7 +233,32 @@ static peripheral_irq_map_t irq_mappings[] = {
 
 #define INTERRUPT_CORE0_CPU_INT_THRESH_REG 0x194
 
-#define MATRIX_REG( reg ) *( (volatile uint32_t *) ( INT_MATRIX_BASE + reg ) )
+static inline void esp32_cpu_int_enable_all( void )
+{
+  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_ENABLE_REG ) = 0xffffffffU;
+}
+
+static inline void esp32_cpu_int_type_level( void )
+{
+  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_TYPE_REG ) = 0x0U;
+}
+
+static inline void esp32_cpu_int_clear( uint8_t cpu_vector )
+{
+  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 1U << cpu_vector;
+  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 0U;
+}
+
+static inline void esp32_cpu_int_set_prio( uint8_t cpu_int, uint32_t prio )
+{
+  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_PRI_n_REG + cpu_int * sizeof( uint32_t ) ) = prio;
+}
+
+static inline uint32_t esp32_cpu_int_get_prio( uint8_t cpu_int )
+{
+  return MATRIX_REG( INTERRUPT_CORE0_CPU_INT_PRI_n_REG + cpu_int * sizeof( uint32_t ) );
+}
+#endif
 
 /*
  * Returned by get_active_interrupt() when no mapped source is asserting.
@@ -163,16 +273,6 @@ static peripheral_irq_map_t irq_mappings[] = {
  */
 #define NO_ACTIVE_VECTOR ( (rtems_vector_number) RISCV_MAXIMUM_EXTERNAL_INTERRUPTS )
 
-static uint64_t get_int_status( void )
-{
-  uint64_t total_set = MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_1_REG );
-
-  total_set <<= 32;
-  total_set |= MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_0_REG );
-
-  return total_set;
-}
-
 static uint8_t int_periph_to_cpu( rtems_vector_number vector )
 {
   size_t map_count = RTEMS_ARRAY_SIZE( irq_mappings );
@@ -188,13 +288,31 @@ static uint8_t int_periph_to_cpu( rtems_vector_number vector )
 
 static rtems_vector_number get_active_interrupt( uint8_t cpu_vector )
 {
-  uint64_t total_set = get_int_status();
+  uint32_t s0 = MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_0_REG );
+  uint32_t s1 = MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_1_REG );
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  uint32_t s2 = MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_2_REG );
+#endif
 
-  for ( rtems_vector_number i = 0; i < sizeof( uint64_t ) * 8; i++ ) {
-    if ( ( total_set & 0x1 ) != 0 && int_periph_to_cpu( i ) == cpu_vector ) {
-      return i;
+  for ( size_t i = 0; i < RTEMS_ARRAY_SIZE( irq_mappings ); i++ ) {
+    if ( irq_mappings[ i ].cpu_int == cpu_vector ) {
+      rtems_vector_number vec = irq_mappings[ i ].peripheral_int;
+      if ( vec < 32 ) {
+        if ( ( s0 & ( 1U << vec ) ) != 0 ) {
+          return vec;
+        }
+      } else if ( vec < 64 ) {
+        if ( ( s1 & ( 1U << ( vec - 32 ) ) ) != 0 ) {
+          return vec;
+        }
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+      } else if ( vec < RISCV_MAXIMUM_EXTERNAL_INTERRUPTS ) {
+        if ( ( s2 & ( 1U << ( vec - 64 ) ) ) != 0 ) {
+          return vec;
+        }
+#endif
+      }
     }
-    total_set >>= 1;
   }
 
   /*
@@ -216,11 +334,6 @@ static uintptr_t periph_int_to_map_reg( rtems_vector_number vector )
   return vector * sizeof( uint32_t );
 }
 
-static uintptr_t cpu_int_to_prio_reg( uint8_t cpu_int )
-{
-  return INTERRUPT_CORE0_CPU_INT_PRI_n_REG + cpu_int * sizeof( uint32_t );
-}
-
 void _RISCV_Interrupt_dispatch( uintptr_t mcause, Per_CPU_Control *cpu_self )
 {
   (void) cpu_self;
@@ -243,8 +356,7 @@ void _RISCV_Interrupt_dispatch( uintptr_t mcause, Per_CPU_Control *cpu_self )
    * handler table.
    */
   if ( active == NO_ACTIVE_VECTOR ) {
-    MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 1U << cpu_vector;
-    MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 0U;
+    esp32_cpu_int_clear( cpu_vector );
     return;
   }
 
@@ -265,9 +377,9 @@ void bsp_interrupt_facility_initialize( void )
   uint32_t cookie = riscv_interrupt_disable();
 
   /* Enable all cpu-side interrupts */
-  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_ENABLE_REG ) = 0xffffffffU;
+  esp32_cpu_int_enable_all();
   /* Set everything to level interrupts */
-  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_TYPE_REG ) = 0x0U;
+  esp32_cpu_int_type_level();
   /*
    * From 0, because source 0 is the WiFi MAC.  The CPU-side channel 0 is the
    * reserved one, and nothing here indexes channels by vector.
@@ -324,31 +436,59 @@ rtems_status_code bsp_interrupt_is_pending(
   bool               *pending
 )
 {
-  uint64_t total_set;
-  uint64_t bit = 1;
-
   bsp_interrupt_assert( bsp_interrupt_is_valid_vector( vector ) );
   bsp_interrupt_assert( pending != NULL );
 
-  total_set = get_int_status();
-  bit <<= vector;
-
-  *pending = ( total_set & bit ) != 0;
+  if ( vector < 32 ) {
+    *pending = ( MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_0_REG ) & ( 1U << vector ) ) != 0;
+  } else if ( vector < 64 ) {
+    *pending = ( MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_1_REG ) & ( 1U << ( vector - 32 ) ) ) != 0;
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  } else if ( vector < RISCV_MAXIMUM_EXTERNAL_INTERRUPTS ) {
+    *pending = ( MATRIX_REG( INTERRUPT_CORE0_INTR_STATUS_2_REG ) & ( 1U << ( vector - 64 ) ) ) != 0;
+#endif
+  } else {
+    *pending = false;
+  }
 
   return RTEMS_SUCCESSFUL;
 }
 
+#if !defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
 #define SYSTEM_CPU_INTR_FROM_CPU_0_REG 0x28
 #define SYSTEM_CPU_INTR_FROM_CPU_1_REG 0x2C
 #define SYSTEM_CPU_INTR_FROM_CPU_2_REG 0x30
 #define SYSTEM_CPU_INTR_FROM_CPU_3_REG 0x34
 
 #define SYSREG_REG( reg ) *( (volatile uint32_t *) ( SYSREG_BASE + reg ) )
+#endif
 
 rtems_status_code bsp_interrupt_raise( rtems_vector_number vector )
 {
   bsp_interrupt_assert( bsp_interrupt_is_valid_vector( vector ) );
 
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  switch ( vector ) {
+    case SW_INTR_0:
+      INTPRI_REG( 0x90 ) = 1;
+      INTPRI_CPU_REG( 0x90 ) = 1;
+      break;
+    case SW_INTR_1:
+      INTPRI_REG( 0x94 ) = 1;
+      INTPRI_CPU_REG( 0x94 ) = 1;
+      break;
+    case SW_INTR_2:
+      INTPRI_REG( 0x98 ) = 1;
+      INTPRI_CPU_REG( 0x98 ) = 1;
+      break;
+    case SW_INTR_3:
+      INTPRI_REG( 0x9c ) = 1;
+      INTPRI_CPU_REG( 0x9c ) = 1;
+      break;
+    default:
+      return RTEMS_UNSATISFIED;
+  }
+#else
   switch ( vector ) {
     case SW_INTR_0:
       SYSREG_REG( SYSTEM_CPU_INTR_FROM_CPU_0_REG ) = 1;
@@ -365,6 +505,7 @@ rtems_status_code bsp_interrupt_raise( rtems_vector_number vector )
     default:
       return RTEMS_UNSATISFIED;
   }
+#endif
 
   return RTEMS_SUCCESSFUL;
 }
@@ -376,11 +517,31 @@ rtems_status_code bsp_interrupt_clear( rtems_vector_number vector )
   bsp_interrupt_assert( bsp_interrupt_is_valid_vector( vector ) );
 
   cpu_int = int_periph_to_cpu( vector );
-
-  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 1U << cpu_int;
-  MATRIX_REG( INTERRUPT_CORE0_CPU_INT_CLEAR_REG ) = 0U;
+  esp32_cpu_int_clear( cpu_int );
 
   /* If this is a software interrupt, clear it. */
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  switch ( vector ) {
+    case SW_INTR_0:
+      INTPRI_REG( 0x90 ) = 0;
+      INTPRI_CPU_REG( 0x90 ) = 0;
+      break;
+    case SW_INTR_1:
+      INTPRI_REG( 0x94 ) = 0;
+      INTPRI_CPU_REG( 0x94 ) = 0;
+      break;
+    case SW_INTR_2:
+      INTPRI_REG( 0x98 ) = 0;
+      INTPRI_CPU_REG( 0x98 ) = 0;
+      break;
+    case SW_INTR_3:
+      INTPRI_REG( 0x9c ) = 0;
+      INTPRI_CPU_REG( 0x9c ) = 0;
+      break;
+    default:
+      break;
+  }
+#else
   switch ( vector ) {
     case SW_INTR_0:
       SYSREG_REG( SYSTEM_CPU_INTR_FROM_CPU_0_REG ) = 0;
@@ -397,6 +558,7 @@ rtems_status_code bsp_interrupt_clear( rtems_vector_number vector )
     default:
       break;
   }
+#endif
 
   return RTEMS_SUCCESSFUL;
 }
@@ -459,8 +621,7 @@ rtems_status_code bsp_interrupt_set_priority(
   bsp_interrupt_assert( bsp_interrupt_is_valid_vector( vector ) );
 
   cpu_int = int_periph_to_cpu( vector );
-
-  MATRIX_REG( cpu_int_to_prio_reg( cpu_int ) ) = priority;
+  esp32_cpu_int_set_prio( cpu_int, priority );
 
   return RTEMS_SUCCESSFUL;
 }
@@ -476,7 +637,7 @@ rtems_status_code bsp_interrupt_get_priority(
   bsp_interrupt_assert( priority != NULL );
 
   cpu_int = int_periph_to_cpu( vector );
-  *priority = MATRIX_REG( cpu_int_to_prio_reg( cpu_int ) );
+  *priority = esp32_cpu_int_get_prio( cpu_int );
 
   /* adjust priority to the range RTEMS uses */
   *priority -= 1;
diff --git a/bsps/riscv/esp32/spi/spi_c3.c b/bsps/riscv/esp32/spi/spi_c3.c
index c4de95c..e6b10ba 100644
--- a/bsps/riscv/esp32/spi/spi_c3.c
+++ b/bsps/riscv/esp32/spi/spi_c3.c
@@ -57,9 +57,15 @@
 #include <inttypes.h>
 #include <string.h>
 
+#ifndef SPI2_BASE
 #define SPI2_BASE 0x60024000U
+#endif
+#ifndef GPIO_BASE
 #define GPIO_BASE 0x60004000U
+#endif
+#ifndef IO_MUX_BASE
 #define IO_MUX_BASE 0x60009000U
+#endif
 
 #define SPI_REG( off ) ( *(volatile uint32_t *) ( SPI2_BASE + (off) ) )
 
diff --git a/bsps/riscv/esp32/start/bspstart.c b/bsps/riscv/esp32/start/bspstart.c
index 991eed9..0d4d649 100644
--- a/bsps/riscv/esp32/start/bspstart.c
+++ b/bsps/riscv/esp32/start/bspstart.c
@@ -80,9 +80,21 @@
 
 void bsp_start( void )
 {
+  uint32_t timg_wdtconfig0;
+
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  /* Unlock and disable LP WDT on C6 */
+  WRITE_REG( LP_WDT_BASE, 0x18, 0x50d83aa1U );
+  uint32_t lp_wdt = READ_REG( LP_WDT_BASE, 0x0 );
+  WRITE_REG( LP_WDT_BASE, 0x0, lp_wdt & ~( BSP_BIT32( 31 ) | BSP_BIT32( 8 ) ) );
+
+  /* Unlock and auto-feed super watchdog (SWD) on C6 */
+  WRITE_REG( LP_WDT_BASE, 0x20, 0x50d83aa1U );
+  uint32_t swd = READ_REG( LP_WDT_BASE, 0x1c );
+  WRITE_REG( LP_WDT_BASE, 0x1c, swd | BSP_BIT32( 18 ) );
+#else
   uint32_t wdt_config0;
   uint32_t swd_conf;
-  uint32_t timg_wdtconfig0;
 
   /* disable RTC watchdog write protection */
   RTC_WRITE( RTC_CNTL_WDTWPROTECT_REG, RTC_CNTL_WDTWPROTECT_KEY );
@@ -120,6 +132,7 @@ void bsp_start( void )
 
   /* clear SWD stall register */
   RTC_WRITE( RTC_CNTL_SW_CPU_STALL_REG, 0 );
+#endif
 
   /*
    * Disable timergroup 0's watchdog, the same correction as above -- and
@@ -168,7 +181,11 @@ BSP_START_TEXT_SECTION static inline void copy_from_flash_offset(
 )
 {
   /* The RAM load sections are offset from 0x0, offset from mapped flash base */
+#if defined(LIBBSP_ESP32_C6_CHIP_DEFINITIONS_H)
+  uintptr_t flash_base = 0x42000000;
+#else
   uintptr_t flash_base = 0x3c000000;
+#endif
 
   uintptr_t flash_address = ( (uintptr_t) src );
   flash_address += flash_base;
diff --git a/spec/build/bsps/riscv/esp32/abi.yml b/spec/build/bsps/riscv/esp32/abi.yml
index 7f773db..05ecc88 100644
--- a/spec/build/bsps/riscv/esp32/abi.yml
+++ b/spec/build/bsps/riscv/esp32/abi.yml
@@ -6,7 +6,16 @@ actions:
 build-type: option
 copyrights:
 - Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
 default:
+- enabled-by: riscv/esp32c3db
+  value:
+  - -march=rv32imc
+  - -mabi=ilp32
+- enabled-by: riscv/esp32c6db
+  value:
+  - -march=rv32imac
+  - -mabi=ilp32
 - enabled-by: true
   value:
   - -march=rv32imc
diff --git a/spec/build/bsps/riscv/esp32/bspesp32c6db.yml b/spec/build/bsps/riscv/esp32/bspesp32c6db.yml
new file mode 100644
index 0000000..90a1b1c
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/bspesp32c6db.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+arch: riscv
+bsp: esp32c6db
+build-type: bsp
+cflags: []
+copyrights:
+- Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+cppflags: []
+enabled-by: true
+family: esp32
+includes: []
+install:
+- destination: ${BSP_INCLUDEDIR}/${ESPRESSIF_CHIP_VARIANT}
+  source:
+  - bsps/riscv/esp32/include/c6/chip_definitions.h
+links:
+- role: build-dependency
+  uid: ../../opto2
+- role: build-dependency
+  uid: grp
+source: []
+type: build
diff --git a/spec/build/bsps/riscv/esp32/grp.yml b/spec/build/bsps/riscv/esp32/grp.yml
index b47bfbb..5fb8e4b 100644
--- a/spec/build/bsps/riscv/esp32/grp.yml
+++ b/spec/build/bsps/riscv/esp32/grp.yml
@@ -40,6 +40,16 @@ links:
   uid: optiramsize
 - role: build-dependency
   uid: optdramromreserve
+- role: build-dependency
+  uid: optramorigin
+- role: build-dependency
+  uid: optramsize
+- role: build-dependency
+  uid: optramcodeorigin
+- role: build-dependency
+  uid: optdataflashmappedbase
+- role: build-dependency
+  uid: optiramdelta
 - role: build-dependency
   uid: optflashsize
 - role: build-dependency
diff --git a/spec/build/bsps/riscv/esp32/linkcmds.yml b/spec/build/bsps/riscv/esp32/linkcmds.yml
index 98281d0..27e6a79 100644
--- a/spec/build/bsps/riscv/esp32/linkcmds.yml
+++ b/spec/build/bsps/riscv/esp32/linkcmds.yml
@@ -8,29 +8,20 @@ content: |
   MEMORY {
     CODE_FLASH_MAPPED : ORIGIN = 0x42000000, LENGTH = ${ESP32C_CODE_REGION_SIZE}
     CODE_FLASH_RAW : ORIGIN = 0x0, LENGTH = ${ESP32C_CODE_REGION_SIZE}
-    DATA_FLASH_MAPPED : ORIGIN = 0x3c000000 + ${ESP32C_CODE_REGION_SIZE}, LENGTH = ${ESP32C_FLASH_SIZE} - ${ESP32C_CODE_REGION_SIZE}
+    DATA_FLASH_MAPPED : ORIGIN = ${ESP32C_DATA_FLASH_MAPPED_BASE} + ${ESP32C_CODE_REGION_SIZE}, LENGTH = ${ESP32C_FLASH_SIZE} - ${ESP32C_CODE_REGION_SIZE}
     DATA_FLASH_RAW : ORIGIN = 0x0 + ${ESP32C_CODE_REGION_SIZE}, LENGTH = ${ESP32C_FLASH_SIZE} - ${ESP32C_CODE_REGION_SIZE}
     /*
-     * One SRAM, two windows.  The part reaches the same bytes at 0x3fc80000
-     * on the data bus and at 0x40380000 on the instruction bus, so a byte
-     * given to RAM_CODE is a byte taken from RAM -- the two regions are carved
-     * out of one block rather than being separate memories, and they must not
-     * overlap or the linker will hand the same storage out twice.
+     * One SRAM, two windows on C3 (unified window on C6).
      *
      * RAM_CODE holds code that must not execute from flash.  With
      * ESP32C_IRAM_REGION_SIZE at its default of zero the region is empty and
      * RAM is the whole SRAM, which is what a configuration that needs none
      * should pay.
      *
-     * The top of the block is not ours either.  The ROM keeps its own data
-     * there, from 0x3fcce710 up, and this BSP calls ROM code continuously --
-     * ets_*, the PHY routines, and the WiFi libraries' ROM halves.  So
-     * ESP32C_DRAM_ROM_RESERVE_SIZE comes off the top for the same reason
-     * ESP32C_IRAM_REGION_SIZE comes off the bottom: it is memory that
-     * exists and is spoken for.
+     * The top of the block is reserved for ROM if ESP32C_DRAM_ROM_RESERVE_SIZE > 0.
      */
-    RAM_CODE : ORIGIN = 0x40380000, LENGTH = ${ESP32C_IRAM_REGION_SIZE}
-    RAM : ORIGIN = 0x3fc80000 + ${ESP32C_IRAM_REGION_SIZE}, LENGTH = 0x50000 - ${ESP32C_IRAM_REGION_SIZE} - ${ESP32C_DRAM_ROM_RESERVE_SIZE}
+    RAM_CODE : ORIGIN = ${ESP32C_RAM_CODE_ORIGIN}, LENGTH = ${ESP32C_IRAM_REGION_SIZE}
+    RAM : ORIGIN = ${ESP32C_RAM_ORIGIN} + ${ESP32C_IRAM_REGION_SIZE}, LENGTH = ${ESP32C_RAM_SIZE} - ${ESP32C_IRAM_REGION_SIZE} - ${ESP32C_DRAM_ROM_RESERVE_SIZE}
   }
 
   /*
@@ -38,7 +29,7 @@ content: |
    * write .fast_text through the window the silicon guarantees for stores.
    * bsp_start_copy_sections() uses it; see bspstart.c.
    */
-  esp32c_iram_to_dram_delta = 0x3fc80000 - 0x40380000;
+  esp32c_iram_to_dram_delta = ${ESP32C_IRAM_TO_DRAM_DELTA};
 
   /*
    * The first address above the region, which is the first address the ROM
@@ -47,7 +38,7 @@ content: |
    * so that a weak reference can tell a BSP that reserves nothing from one
    * that reserves a different amount.  A literal can do neither.
    */
-  esp32c_dram_rom_reserve_begin = 0x3fc80000 + 0x50000 - ${ESP32C_DRAM_ROM_RESERVE_SIZE};
+  esp32c_dram_rom_reserve_begin = ${ESP32C_RAM_ORIGIN} + ${ESP32C_RAM_SIZE} - ${ESP32C_DRAM_ROM_RESERVE_SIZE};
 
   REGION_ALIAS ("REGION_START", CODE_FLASH_MAPPED);
   REGION_ALIAS ("REGION_START_LOAD", CODE_FLASH_RAW);
diff --git a/spec/build/bsps/riscv/esp32/optchipvariant.yml b/spec/build/bsps/riscv/esp32/optchipvariant.yml
index 8d497df..d893bcf 100644
--- a/spec/build/bsps/riscv/esp32/optchipvariant.yml
+++ b/spec/build/bsps/riscv/esp32/optchipvariant.yml
@@ -5,9 +5,12 @@ actions:
 build-type: option
 copyrights:
 - Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
 default:
 - enabled-by: riscv/esp32c3db
   value: 'c3'
+- enabled-by: riscv/esp32c6db
+  value: 'c6'
 - enabled-by: true
   value: 'invalid'
 description: |
diff --git a/spec/build/bsps/riscv/esp32/optdataflashmappedbase.yml b/spec/build/bsps/riscv/esp32/optdataflashmappedbase.yml
new file mode 100644
index 0000000..7a80497
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/optdataflashmappedbase.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+actions:
+- get-integer: null
+- assert-uint32: null
+- env-assign: null
+- format-and-define: null
+build-type: option
+copyrights:
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+default:
+- enabled-by: riscv/esp32c3db
+  value: 0x3c000000
+- enabled-by: riscv/esp32c6db
+  value: 0x42000000
+- enabled-by: true
+  value: 0x3c000000
+description: |
+  Base address for memory-mapped flash read-only data.
+enabled-by: true
+format: '{:#010x}'
+links: []
+name: ESP32C_DATA_FLASH_MAPPED_BASE
+type: build
diff --git a/spec/build/bsps/riscv/esp32/optdramromreserve.yml b/spec/build/bsps/riscv/esp32/optdramromreserve.yml
index ec813e0..71fd8ca 100644
--- a/spec/build/bsps/riscv/esp32/optdramromreserve.yml
+++ b/spec/build/bsps/riscv/esp32/optdramromreserve.yml
@@ -8,6 +8,8 @@ build-type: option
 copyrights:
 - Copyright (C) 2026 Samuel Price <thesamprice@gmail.com>
 default:
+- enabled-by: riscv/esp32c6db
+  value: 0x16000
 - enabled-by: true
   value: 0x0
 description: |
diff --git a/spec/build/bsps/riscv/esp32/optiramdelta.yml b/spec/build/bsps/riscv/esp32/optiramdelta.yml
new file mode 100644
index 0000000..c437957
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/optiramdelta.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+actions:
+- get-string: null
+- env-assign: null
+- format-and-define: null
+build-type: option
+copyrights:
+- Copyright (C) 2026 Samuel Price <thesamprice@gmail.com>
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+default:
+- enabled-by: riscv/esp32c3db
+  value: '0x3fc80000 - 0x40380000'
+- enabled-by: riscv/esp32c6db
+  value: '0'
+- enabled-by: true
+  value: '0'
+description: |
+  Distance from instruction window to data window for SRAM.
+enabled-by: true
+format: '{}'
+links: []
+name: ESP32C_IRAM_TO_DRAM_DELTA
+type: build
diff --git a/spec/build/bsps/riscv/esp32/optramcodeorigin.yml b/spec/build/bsps/riscv/esp32/optramcodeorigin.yml
new file mode 100644
index 0000000..3ffb786
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/optramcodeorigin.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+actions:
+- get-integer: null
+- assert-uint32: null
+- env-assign: null
+- format-and-define: null
+build-type: option
+copyrights:
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+default:
+- enabled-by: riscv/esp32c3db
+  value: 0x40380000
+- enabled-by: riscv/esp32c6db
+  value: 0x40800000
+- enabled-by: true
+  value: 0x40380000
+description: |
+  Internal RAM code execution base address.
+enabled-by: true
+format: '{:#010x}'
+links: []
+name: ESP32C_RAM_CODE_ORIGIN
+type: build
diff --git a/spec/build/bsps/riscv/esp32/optramorigin.yml b/spec/build/bsps/riscv/esp32/optramorigin.yml
new file mode 100644
index 0000000..1fbd63e
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/optramorigin.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+actions:
+- get-integer: null
+- assert-uint32: null
+- env-assign: null
+- format-and-define: null
+build-type: option
+copyrights:
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+default:
+- enabled-by: riscv/esp32c3db
+  value: 0x3fc80000
+- enabled-by: riscv/esp32c6db
+  value: 0x40800000
+- enabled-by: true
+  value: 0x3fc80000
+description: |
+  Internal RAM data base address.
+enabled-by: true
+format: '{:#010x}'
+links: []
+name: ESP32C_RAM_ORIGIN
+type: build
diff --git a/spec/build/bsps/riscv/esp32/optramsize.yml b/spec/build/bsps/riscv/esp32/optramsize.yml
new file mode 100644
index 0000000..caa29cb
--- /dev/null
+++ b/spec/build/bsps/riscv/esp32/optramsize.yml
@@ -0,0 +1,23 @@
+SPDX-License-Identifier: CC-BY-SA-4.0 OR BSD-2-Clause
+actions:
+- get-integer: null
+- assert-uint32: null
+- env-assign: null
+- format-and-define: null
+build-type: option
+copyrights:
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
+default:
+- enabled-by: riscv/esp32c3db
+  value: 0x50000
+- enabled-by: riscv/esp32c6db
+  value: 0x80000
+- enabled-by: true
+  value: 0x50000
+description: |
+  Internal RAM total size in bytes.
+enabled-by: true
+format: '{:#010x}'
+links: []
+name: ESP32C_RAM_SIZE
+type: build
diff --git a/spec/build/bsps/riscv/esp32/optvariant.yml b/spec/build/bsps/riscv/esp32/optvariant.yml
index 0c4d394..ddb2544 100644
--- a/spec/build/bsps/riscv/esp32/optvariant.yml
+++ b/spec/build/bsps/riscv/esp32/optvariant.yml
@@ -6,9 +6,12 @@ actions:
 build-type: option
 copyrights:
 - Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
+- Copyright (C) 2026 Peter M. <petermm@gmail.com>
 default:
 - enabled-by: riscv/esp32c3db
   value: 'c3/chip_definitions.h'
+- enabled-by: riscv/esp32c6db
+  value: 'c6/chip_definitions.h'
 - enabled-by: true
   value: 'invalid'
 description: ''
diff --git a/spec/build/bsps/riscv/optbspstartcopysections.yml b/spec/build/bsps/riscv/optbspstartcopysections.yml
index fb7a462..debd719 100644
--- a/spec/build/bsps/riscv/optbspstartcopysections.yml
+++ b/spec/build/bsps/riscv/optbspstartcopysections.yml
@@ -7,7 +7,9 @@ build-type: option
 copyrights:
 - Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
 default:
-- enabled-by: riscv/esp32c3db
+- enabled-by:
+  - riscv/esp32c3db
+  - riscv/esp32c6db
   value: true
 - enabled-by: true
   value: false
diff --git a/spec/build/bsps/riscv/optespressifdirectboot.yml b/spec/build/bsps/riscv/optespressifdirectboot.yml
index 7d62a80..ee32ac4 100644
--- a/spec/build/bsps/riscv/optespressifdirectboot.yml
+++ b/spec/build/bsps/riscv/optespressifdirectboot.yml
@@ -7,7 +7,9 @@ build-type: option
 copyrights:
 - Copyright (C) 2026 Kinsey Moore <wkmoore@gmail.com>
 default:
-- enabled-by: riscv/esp32c3db
+- enabled-by:
+  - riscv/esp32c3db
+  - riscv/esp32c6db
   value: true
 - enabled-by: true
   value: false
-- 
2.50.1 (Apple Git-155)

